### ISSUE TITLE
Switcheroo variables passed to decorator

### ISSUE THREAD

**Issue by terrdavis** [beartype/beartype#1]
For full-fat type checking of arbitrary length containers, instead of providing wrapped built in types that must be imported, wrap them at runtime and swap them out in place. Maybe this is only possible in cpython and relies on implementation details?

**Comment by leycec** [comment]
**Sweetness.** I love this sort of abstract rumination. As the physical world melts down around us, let us bravely cogitate!

> For full-fat type checking of arbitrary length containers, ...wrap them at runtime and swap them out in place.

↖ Excellent idea here ↖

In slightly higher-level languages like Ruby and JavaScript, directly monkey-patching builtin types is trivial, safe, and commonly done by various popular frameworks (e.g., [RSpec in Ruby](https://rspec.info)).

Meanwhile, this is Python. CPython itself prohibits directly monkey-patching builtin types:

```
>>> list.append = lambda self, obj: self.append(obj + ' uhoh')                                                       
TypeError: can't set attributes of built-in/extension type 'list'
```

As you suggest, this is circumventable by replacing builtin types with comparable subclasses via the stdlib `builtins` module:

```
>>> import builtins
>>> class UnicornList(list):
>>>     def append(self, obj): super().append(obj + ' ohnoudint')
>>> builtins.list = UnicornList
>>> crabrave_list = list(('dis', 'is', 'balls'))
>>> crabrave_list.append('uwutm8')
>>> print(crabrave_list)
['dis', 'is', 'balls', 'uwutm8 ohnoudint']
```

So what's the [Gordian knot](https://en.wikipedia.org/wiki/Gordian_Knot) here?

**Syntactic sugar.** Lists typically aren't explicitly instantiated via their builtin type name; they're instantiated via `[`- and `]`-delimited syntactic sugar. For unknown reasons that exceed my meagre intellect, the above approach does *not* apply to lists instantiated via the standard approach:

```
>>> crabrave_list = ['dis', 'is', 'balls']
>>> crabrave_list.append('uwutm8')
>>> print(crabrave_list)
['dis', 'is', 'balls', 'uwutm8']
```

Since a solution that only solves the least likely of two possible cases is no solution, our hands are tied by CPython internals implemented well before even more important stuff like Doom and GeoCities and poorly maintained GeoCities sites about Doom were first created.

> Maybe this is only possible in cpython and relies on implementation details?

Yes. Yes, it does.

Technically, a [risky third-party library named `forbiddenfruit` that no one is supposed to know about or ever use](https://github.com/clarete/forbiddenfruit) *does* exist. Pragmatically, this library [commonly causes fatal runtime segfaults](https://github.com/clarete/forbiddenfruit/issues?q=is%3Aissue+is%3Aopen+segfault), [requires use of a function called `curse`](https://github.com/clarete/forbiddenfruit#tiny-example), [appears to have only been tested up to Python 3.3](https://github.com/clarete/forbiddenfruit#compatibility), and (in any case) only applies to CPython.

But let's leave this issue open. I'd love to find the Shangri La of builtin monkey-patching techniques that would safely, portably, and efficiently enable us to transparently "improve" core container types across Python interpreters. Someone who is smarter than me: *help us and all the Internet karma will be yours.*

**Comment by leycec** [comment]
And [let's not forget this](https://stackoverflow.com/questions/38257613/how-to-monkey-patch-python-list-setitem-method):

> Even if you got all of this right, the C level code that directly calls the `PyList_*` or `PySequence_Fast*` APIs bypasses lookup of `__setitem__`/`mp_ass_subscript`/`sq_ass_item` entirely, so your code wouldn't be invoked. – ShadowRanger

Thanks, [ShadowRanger](https://stackoverflow.com/users/364696/shadowranger), for destroying our collective hopes and dreams.

#### tl;dr

**can't. be. done.** <sup>probably.</sup>

**Comment by leycec** [comment]
That said, I *just* ideated a few helpful heuristics for deeply type-checking a few common types of containers – all with amortized time complexity `O(1)`.

## Hashability Is Deep Immutability... Mostly

The core insight here is that there effectively exists a [one-to-one mapping between hashability and deep immutability in Python](https://inventwithpython.com/blog/2019/02/01/hashable-objects-must-be-immutable), excluding the well-known caveat of user-defined mutable classes whose default `__hash__()` implementations return their instance IDs. This caveat is trivially detectable and thus ignorable for most purposes.

Why is this useful? Because **we only need to type-check deeply immutable objects once.** Makes sense, it does. Note the meaningful distinction between deep and shallow immutability, however; a tuple is guaranteed to be shallowly but *not* deeply immutable. This means that we walk softly and carry a big stick.

Although there exists no means of directly deciding whether an object is deeply immutable in Python, there *does* exist a means of directly deciding whether an object is hashable in Python: e.g.,

```
obj = ('TNT', ["it's", "dynamite"])
is_obj_hashable = False
try:
    hash(obj)
    is_obj_hashable = True
except:
    pass
```

Although tuples are of course shallowly immutable, tuples containing mutable objects are *not* hashable and thus *not* deeply immutable. The above example demonstrates this by defining a tuple containing an immutable object (i.e., a string) and a mutable object (i.e., a list) and then attempting to hash that tuple, which raises a `TypeError` that we conveniently ignore.

The above snippet successfully detects hashability and thus deep immutability for *builtin* types, but yields false positives for user-defined mutable types. There exist three cases here:

* The **common case,** in which a user-defined class fails to override the `__hash__()` dunder method and thus inherits that method from `object`, whose default implementation of that method simply returns that object's ID.
* The **uncommon non-erroneous case,** in which a user-defined class correctly overrides that method to either:
  * If instances of that class are mutable, return `None`. This is the [standard idiom to indicate unhashability in Python](https://help.semmle.com/wiki/display/PYTHON/Non-standard+exception+raised+in+special+method).
  * If instances of that class are immutable, return something non-`None`.
* The **uncommon erroneous case,** in which a user-defined mutable class incorrectly overrides that method to return something non-`None`. That's bad, but there's mostly nothing we can do about that. User errors are on the user. PEBKAC is our sole defence, bro!

The common case case *must* be explicitly guarded against. The uncommon non-erroneous case needn't be guarded against, because it behaves as expected. The uncommon non-erroneous case needn't be guarded against either, because it's a low-level error beyond our control or capacity to detect.

Thankfully, guarding against the common case is easy. Just exclude types whose `__hash__()` dunder method is *exactly* the `object.__hash__()` dunder method: e.g.,

```
class LetThereBeBeartype(object): pass
obj = LetThereBeBeartype
is_obj_hashable = False
try:
    hash(obj)
    is_obj_hashable = type(obj).__hash__ is not object.__hash__
except:
    pass
```

This extension now accurately detects hashability and thus deep immutability. There are performance concerns even here, however:

* **Hashing time complexity.** Although the builtin `hash()` function is known to be reasonably fast, this only applies to [builtin scalar types that internally cache their hashes](https://stackoverflow.com/a/3787481/2809027) (e.g., `str`, `int`). The `tuple.__hash__()` implementation is a notable counterexample; to reduce space complexity, this implementation intentionally does *not* internally cache its hash. Since this implementation's time complexity is `O(len(tuple))`, naively hashing arbitrary objects passed to and returned from `@beartype`-decorated callables is a BadIdea™.
* **Hashing space complexity.** The prior concern can, of course, be ameliorated by an internal `@beartype` hash cache mapping from object ID to hash. Builtin scalar types already internally cache hashes and would thus be omitted from this cache. For efficiency, *all* other objects would need to be cached. The space complexity required to do so could rapidly become ugly. That concern can, of course, also be ameliorated by limiting the size of that cache in an intelligent manner – like, say, with the Least Recently Used (LRU) replacement strategy implemented by the builtin `@functools.lru_cache` decorator, which [appears to have reasonable performance](https://www.cameronmacleod.com/blog/python-lru-cache). If that fails, there's also [this clever LRU dictionary implementation](https://stackoverflow.com/a/8334739/2809027).

*Phew!* So, hashing time and space complexity is a demonstrable concern – but only trivially so, because we can hack around those concerns with sufficient foresight, which we have.

## ...So Good, So What?

By the above argument, we see that we can efficiently decide whether arbitrary objects are deeply immutable or not. That's good, because **deeply immutable objects only need to be type-checked once.**

Here's how we do that:

1. Add a new private `beartype._decor.cache` submodule declaring at least the following functions:
   * `is_object_immutable_deep()`, a `@functools.lru_cache`-decorated function accepting an arbitrary object and returning `True` if and only if that object is deeply immutable by the above algorithm. Since caching object ID <-> boolean pairings has trivial space complexity, this LRU cache should be stupidly large.
   * `beartype_object_immutable_deep()`, a `@functools.lru_cache`-decorated function accepting a deeply immutable object *and* associated type hint and unconditionally returning `None` after deeply type-checking that object. Thanks to LRU caching, this *should* have the intended effect of only deeply type-checking that object on the first call to that function with that object. Since caching object ID <-> `None` pairings has even more trivial space complexity, this LRU cache should be even stupidly larger than that of `is_object_immutable_deep()`.
1. In the `@beartype` decorator implementation:
   1. For each deeply immutable object type-hinted with PEP 484-compliant deep hinting (e.g., `typing.Tuple[str]`) either passed to or returned from a `@beartype`-decorated callable: <sup>*Note that all keys of all mappings are guaranteed to be of immutable types, so we needn't test them.*</sup>
   1. If `is_object_immutable_deep()` returns `True`, unconditionally call `beartype_object_immutable_deep()`.

Of course, we can probably optimize a bit of that heuristic way – in theory, anyway. Rather than the `is_object_immutable_deep()` function, what we *really* want is a function that decides whether an arbitrary PEP 484-compliant type hint describes a deeply immutable type or not. `typing.Tuple[str]` does, for example, but both `typing.Tuple[List[str]]` and `typing.List[str]` do *not*. While the `is_object_immutable_deep()` function is trivially implemented (as above), deciding the deep immutability of `typing` types is probably *not* trivial implemented. There's no well-defined ABC or `Protocol` governing immutability or mutability, so it's unclear exactly how we would generalize that. In any case, `is_object_immutable_deep()` probably suffices for now.

The above then gives us amortized time complexity `O(1)` for deeply type-checking *all* deeply immutable types. That sounds nice, doesn't it? Since deeply immutable types probably constitute a reasonable fraction of all container types, we're well on our way to type-checking first Berlin, then Manhattan, and then... *the World!* <sup>muhah–cough, wheeze</sup>

**Comment by leycec** [comment]
**[beartype 0.1.0 is now live (and maybe worky) on PyPi](https://pypi.org/project/beartype/0.1.0),** much to the astonishment of my weary brainpan: :sleepy: 

    pip3 install beartype

**Comment by leycec** [comment]
@terrdavis: I *definitely* missed [this commentary](https://gist.github.com/leycec/3d48cc60e7d5fe8860b077acb067dc54#gistcomment-2868908) on my earlier speedread of [your Gist](https://gist.github.com/terrdavis/1b23b7ff8023f55f627199b09cfa6b24):

> It tests for the new error `"TypeError: Subscripted generics cannot be used with class and instance checks"` by always attempting to use the `__origin__` attribute of Generic types for instance checks. I also did some refactoring.

That's super-exciting stuff, bro. When I get the chance, I'd love to integrate that into the upcoming beartype 0.1.1 release – I think, anyway. Volunteer time is scarce for me at the moment, but all of that sounds wondrous. Of course, pull requests are always welcome as well... :grinning:

**Comment by leycec** [comment]
**Note to self:** [related commentary from several months ago at @terrdavis' awesome gist fork](https://gist.github.com/terrdavis/1b23b7ff8023f55f627199b09cfa6b24#gistcomment-3237209), mostly pertaining to randomly selected type-checking of deeply nested containers. *It is fun.*

**Comment by leycec** [comment]
And so it came to pass that in the third day of the ninth month of the Year of Our COVID this issue was summarily terminated, nobody having even the foggiest of a glimmer of an inkling of a cogent idea about how to sanely resolve it.

Actually... *that's a bald-faced lie.* @albanie, I believe it was, had a brilliantly simple solution that [I've since toyed around with and come to see the value-added benefit of](https://github.com/beartype/beartype#nobody-believes-you): **just randomly type-check one item from each container on each call.** Assuming a sufficient number of calls, this heuristic *eventually* type-checks all items from each container with [hard statistical guarantees inspired by the coupon collector's problem](https://en.wikipedia.org/wiki/Coupon_collector%27s_problem). Basically, you need `n logn` calls repeatedly passed the same container to type-check all items from that container – on average, anyway. What won't those credentialed statisticians think up next?

So, this *might* not necessarily be the unsolvable issue we thought it once was. Or maybe it still is. Feel to free to reopen this if you're suddenly assaulted by some clever hackery in the middle of the night. Until then, this is...

**_closed for business_** :door:
